### PR TITLE
add lvis support

### PR DESCRIPTION
**Pull request by xvjiarui**
This PR added LVIS dataset support. 
1. Added LVIS training config.
2. Replace official `coco-api` with the forked one [here](https://github.com/open-mmlab/cocoapi).

### CODE REVIEW

File: mmdet/datasets/dataset_api.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
+from lvis import LVIS
+from pycocotools.coco import COCO
+
+
+class API:
```
Comment (hellock) [inline mmdet/datasets/dataset_api.py]: Maybe renamed to `CocoAPI`.

File: requirements/runtime.txt [hunk 2]
```diff
@@ -1,3 +1,4 @@
+lvis
```
Comment (hellock) [inline requirements/runtime.txt]: Make it optional.

File: mmdet/datasets/__init__.py [hunk 3]
```diff
@@ -2,16 +2,28 @@
...
+    'CityscapesDataset',
+    'GroupSampler',
+    'DistributedGroupSampler',
+    'build_dataloader',
+    'ConcatDataset',
+    'RepeatDataset',
+    'RepeatFactorDataset',
+    'WIDERFaceDataset',
+    'DATASETS',
+    'build_dataset',
```
Comment (yhcao6) [inline mmdet/datasets/__init__.py]: Remove the last comma

File: mmdet/datasets/dataset_wrappers.py [hunk 4]
```diff
@@ -53,3 +56,65 @@ def __getitem__(self, idx):
 
     def __len__(self):
         return self.times * self._ori_len
+
+
+# Modified from https://github.com/facebookresearch/detectron2/blob/41d475b75a230221e21d9cac5d69655e3415e3a4/detectron2/data/samplers/distributed_sampler.py#L57 # noqa
+@DATASETS.register_module
+class RepeatFactorDataset(object):
```
Comment (yhcao6) [inline mmdet/datasets/dataset_wrappers.py]: Add Doc.

File: mmdet/datasets/dataset_wrappers.py [hunk 5]
```diff
@@ -53,3 +56,65 @@ def __getitem__(self, idx):
...
+    def __init__(self, dataset, repeat_thr):
+        self.dataset = dataset
+        self.repeat_thr = repeat_thr
+        self.CLASSES = dataset.CLASSES
+
+        repeat_factors = self._get_repeat_factors(dataset, repeat_thr)
+        repeat_indices = []
+        for dataset_index, repeat_factor in enumerate(repeat_factors):
+            repeat_indices.extend([dataset_index] *
+                                  int(math.ceil(repeat_factor)))
```
Comment (yhcao6) [inline mmdet/datasets/dataset_wrappers.py]: int is redundant.

File: mmdet/datasets/lvis.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+        metrics = metric if isinstance(metric, list) else [metric]
+        allowed_metrics = ['bbox', 'segm', 'proposal', 'proposal_fast']
+        for metric in metrics:
+            if metric not in allowed_metrics:
+                raise KeyError('metric {} is not supported'.format(metric))
+
+        if jsonfile_prefix is None:
+            tmp_dir = tempfile.TemporaryDirectory()
+            jsonfile_prefix = osp.join(tmp_dir.name, 'results')
```
Comment (yhcao6) [inline mmdet/datasets/lvis.py]:
```
else:
    tmp_dir = None
```

File: configs/lvis/mask_rcnn_r50_fpn_sample1e-3_mstrain_2x_lvis.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+data = dict(
+    train=dict(
+        _delete_=True,
+        type='ClassBalancedDataset',
+        oversample_thr=1e-3,
+        dataset=dict(
+            type=dataset_type,
+            ann_file=data_root + 'annotations/lvis_v0.5_train.json',
+            img_prefix=data_root + 'train2017/',
+            pipeline=train_pipeline)))
```
Comment (hellock) [inline configs/lvis/mask_rcnn_r50_fpn_sample1e-3_mstrain_2x_lvis.py:42]: Maybe we can add lvis to `_base_/datasets`

File: mmdet/datasets/lvis.py [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+import numpy as np
+from mmcv.utils import print_log
+from terminaltables import AsciiTable
+
+from .builder import DATASETS
+from .coco import CocoDataset
+
+
+@DATASETS.register_module()
+class LvisDataset(CocoDataset):
```
Comment (hellock) [inline mmdet/datasets/lvis.py:15]: LVISDataset is more preferred.

File: mmdet/datasets/lvis.py [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+                        key = '{}_{}'.format(metric, k)
+                        val = float('{:.3f}'.format(float(v)))
+                        eval_results[key] = val
+                ap_summary = ' '.join([
+                    '{}:{:.3f}'.format(k, float(v))
+                    for k, v in lvis_results.items() if k.startswith('AP')
+                ])
+                eval_results['{}_mAP_copypaste'.format(metric)] = ap_summary
+            lvis_eval.print_results()
+        if tmp_dir is None:
```
Comment (hellock) [inline mmdet/datasets/lvis.py:427]: `if tmp_dir is not None`?

**Pull request merged**